File: output/typescript/tests/08_hsm.ts
class HSMForwardFrameEvent {
    public _message: string;
    public _parameters: Record<string, any> | null;

    constructor(message: string, parameters: Record<string, any> | null) {
        this._message = message;
        this._parameters = parameters;
    }
}


class HSMForwardFrameContext {
    public event: HSMForwardFrameEvent;
    public _return: any;
    public _data: Record<string, any>;

    constructor(event: HSMForwardFrameEvent, default_return: any) {
        this.event = event;
        this._return = default_return;
        this._data = {  };
    }
}


class HSMForwardCompartment {
    public state: string;
    public state_args: Record<string, any>;
    public state_vars: Record<string, any>;
    public enter_args: Record<string, any>;
    public exit_args: Record<string, any>;
    public forward_event: any;
    public parent_compartment: HSMForwardCompartment | null;

    constructor(state: string, parent_compartment: HSMForwardCompartment | null = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    public copy(): HSMForwardCompartment {
        const c = new HSMForwardCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


class HSMForward {
    private _state_stack: Array<any>;
    private __compartment: HSMForwardCompartment;
    private __next_compartment: HSMForwardCompartment | null;
    private _context_stack: Array<any>;
    private log: string[] =     [];

    constructor() {
        this._state_stack = [];
        this._context_stack = [];
        this.log =         [];
        this.__compartment = new HSMForwardCompartment("Child");
        this.__next_compartment = null;
        const __frame_event = new HSMForwardFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    private __kernel(__e: HSMForwardFrameEvent) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new HSMForwardFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new HSMForwardFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new HSMForwardFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    private __router(__e: HSMForwardFrameEvent) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = (this as any)[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    private __transition(next_compartment: HSMForwardCompartment) {
        this.__next_compartment = next_compartment;
    }

    public event_a() {
        const __e = new HSMForwardFrameEvent("event_a", null);
        const __ctx = new HSMForwardFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    public event_b() {
        const __e = new HSMForwardFrameEvent("event_b", null);
        const __ctx = new HSMForwardFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    public get_log(): string[] {
        const __e = new HSMForwardFrameEvent("get_log", null);
        const __ctx = new HSMForwardFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()!._return;
    }

    private _state_Child(__e: HSMForwardFrameEvent) {
        if (__e._message === "event_a") {
            this.log.push("Child:event_a");
        } else if (__e._message === "event_b") {
            this.log.push("Child:event_b_forward");
            this._state_Parent(__e);
        } else if (__e._message === "get_log") {
            this._context_stack[this._context_stack.length - 1]._return = this.log;
            return;;
        }
    }

    private _state_Parent(__e: HSMForwardFrameEvent) {
        if (__e._message === "event_a") {
            this.log.push("Parent:event_a");
        } else if (__e._message === "event_b") {
            this.log.push("Parent:event_b");
        } else if (__e._message === "get_log") {
            this._context_stack[this._context_stack.length - 1]._return = this.log;
            return;;
        }
    }
}


function main() {
    console.log("=== Test 08: HSM Forward ===");
    const s = new HSMForward();

    // event_a should be handled by Child (no forward)
    s.event_a();
    let log = s.get_log();
    if (!log.includes("Child:event_a")) {
        throw new Error(`Expected 'Child:event_a' in log, got ${JSON.stringify(log)}`);
    }
    console.log(`After event_a: ${JSON.stringify(log)}`);

    // event_b should forward to Parent
    s.event_b();
    log = s.get_log();
    if (!log.includes("Child:event_b_forward")) {
        throw new Error(`Expected 'Child:event_b_forward' in log, got ${JSON.stringify(log)}`);
    }
    if (!log.includes("Parent:event_b")) {
        throw new Error(`Expected 'Parent:event_b' in log (forwarded), got ${JSON.stringify(log)}`);
    }
    console.log(`After event_b (forwarded): ${JSON.stringify(log)}`);

    console.log("PASS: HSM forward works correctly");
}

main();

File: output/rust/tests/08_hsm.rs
use std::collections::HashMap;

struct HSMForwardFrameEvent {
    message: String,
    parameters: std::collections::HashMap<String, Box<dyn std::any::Any>>,
}

impl Clone for HSMForwardFrameEvent {
    fn clone(&self) -> Self {
        Self {
            message: self.message.clone(),
            parameters: std::collections::HashMap::new(),
        }
    }
}

impl HSMForwardFrameEvent {
    fn new(message: &str) -> Self {
        Self {
            message: message.to_string(),
            parameters: std::collections::HashMap::new(),
        }
    }
}

struct HSMForwardFrameContext {
    event: HSMForwardFrameEvent,
    _return: Option<Box<dyn std::any::Any>>,
    _data: std::collections::HashMap<String, Box<dyn std::any::Any>>,
}

impl HSMForwardFrameContext {
    fn new(event: HSMForwardFrameEvent, default_return: Option<Box<dyn std::any::Any>>) -> Self {
        Self {
            event,
            _return: default_return,
            _data: std::collections::HashMap::new(),
        }
    }
}

#[derive(Clone)]
struct HSMForwardCompartment {
    state: String,
    state_vars: std::collections::HashMap<String, i32>,
    forward_event: Option<HSMForwardFrameEvent>,
}

impl HSMForwardCompartment {
    fn new(state: &str) -> Self {
        Self {
            state: state.to_string(),
            state_vars: std::collections::HashMap::new(),
            forward_event: None,
        }
    }
}

#[derive(Clone)]
enum HSMForwardStateContext {
    Child,
    Parent,
    Empty,
}

impl Default for HSMForwardStateContext {
    fn default() -> Self {
        HSMForwardStateContext::Child
    }
}

pub struct HSMForward {
    _state_stack: Vec<(String, HSMForwardStateContext)>,
    __compartment: HSMForwardCompartment,
    __next_compartment: Option<HSMForwardCompartment>,
    _context_stack: Vec<HSMForwardFrameContext>,
    log: Vec<String>,
}

impl HSMForward {
    pub fn new() -> Self {
        let mut this = Self {
            _state_stack: vec![],
            _context_stack: vec![],
            log: Vec::new(),
            __compartment: HSMForwardCompartment::new("Child"),
            __next_compartment: None,
        };
let __frame_event = HSMForwardFrameEvent::new("$>");
this.__kernel(__frame_event);
        this
    }

    fn __kernel(&mut self, __e: HSMForwardFrameEvent) {
// Route event to current state
self.__router(&__e);
// Process any pending transition
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    // Exit current state
    let exit_event = HSMForwardFrameEvent::new("<$");
    self.__router(&exit_event);
    // Switch to new compartment
    self.__compartment = next_compartment;
    // Enter new state (or forward event)
    if self.__compartment.forward_event.is_none() {
        let enter_event = HSMForwardFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        // Forward event to new state
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            // Forwarding enter event - just send it
            self.__router(&forward_event);
        } else {
            // Forwarding other event - send $> first, then forward
            let enter_event = HSMForwardFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
    }

    fn __router(&mut self, __e: &HSMForwardFrameEvent) {
match self.__compartment.state.as_str() {
    "Child" => self._state_Child(__e),
    "Parent" => self._state_Parent(__e),
    _ => {}
}
    }

    fn __transition(&mut self, next_compartment: HSMForwardCompartment) {
self.__next_compartment = Some(next_compartment);
    }

    fn _state_stack_push(&mut self) {
let state_context = match self.__compartment.state.as_str() {
    "Child" => HSMForwardStateContext::Child,
    "Parent" => HSMForwardStateContext::Parent,
    _ => HSMForwardStateContext::Empty,
};
self._state_stack.push((self.__compartment.state.clone(), state_context));
    }

    fn _state_stack_pop(&mut self) {
let (saved_state, state_context) = self._state_stack.pop().unwrap();
let exit_event = HSMForwardFrameEvent::new("$<");
self.__router(&exit_event);
self.__compartment.state = saved_state;
match state_context {
    HSMForwardStateContext::Child => {}
    HSMForwardStateContext::Parent => {}
    HSMForwardStateContext::Empty => {}
}
    }

    pub fn event_a(&mut self) {
let mut __e = HSMForwardFrameEvent::new("event_a");
let __ctx = HSMForwardFrameContext::new(__e.clone(), None);
self._context_stack.push(__ctx);
match self.__compartment.state.as_str() {
            "Child" => { self._s_Child_event_a(&__e); }
            "Parent" => { self._s_Parent_event_a(&__e); }
            _ => {}
        }
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    let exit_event = HSMForwardFrameEvent::new("<$");
    self.__router(&exit_event);
    self.__compartment = next_compartment;
    if self.__compartment.forward_event.is_none() {
        let enter_event = HSMForwardFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            self.__router(&forward_event);
        } else {
            let enter_event = HSMForwardFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
self._context_stack.pop();
    }

    pub fn event_b(&mut self) {
let mut __e = HSMForwardFrameEvent::new("event_b");
let __ctx = HSMForwardFrameContext::new(__e.clone(), None);
self._context_stack.push(__ctx);
match self.__compartment.state.as_str() {
            "Child" => { self._s_Child_event_b(&__e); }
            "Parent" => { self._s_Parent_event_b(&__e); }
            _ => {}
        }
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    let exit_event = HSMForwardFrameEvent::new("<$");
    self.__router(&exit_event);
    self.__compartment = next_compartment;
    if self.__compartment.forward_event.is_none() {
        let enter_event = HSMForwardFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            self.__router(&forward_event);
        } else {
            let enter_event = HSMForwardFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
self._context_stack.pop();
    }

    pub fn get_log(&mut self) -> Vec<String> {
let mut __e = HSMForwardFrameEvent::new("get_log");
let __ctx = HSMForwardFrameContext::new(__e.clone(), None);
self._context_stack.push(__ctx);
match self.__compartment.state.as_str() {
            "Child" => { self._s_Child_get_log(&__e); }
            "Parent" => { self._s_Parent_get_log(&__e); }
            _ => {}
        }
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    let exit_event = HSMForwardFrameEvent::new("<$");
    self.__router(&exit_event);
    self.__compartment = next_compartment;
    if self.__compartment.forward_event.is_none() {
        let enter_event = HSMForwardFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            self.__router(&forward_event);
        } else {
            let enter_event = HSMForwardFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
let __ctx = self._context_stack.pop().unwrap();
if let Some(ret) = __ctx._return {
    *ret.downcast::<Vec<String>>().unwrap()
} else {
    Default::default()
}
    }

    fn _state_Child(&mut self, __e: &HSMForwardFrameEvent) {
match __e.message.as_str() {
    "event_a" => { self._s_Child_event_a(__e); }
    "event_b" => { self._s_Child_event_b(__e); }
    "get_log" => { self._s_Child_get_log(__e); }
    _ => self._state_Parent(__e),
}
    }

    fn _state_Parent(&mut self, __e: &HSMForwardFrameEvent) {
match __e.message.as_str() {
    "event_a" => { self._s_Parent_event_a(__e); }
    "event_b" => { self._s_Parent_event_b(__e); }
    "get_log" => { self._s_Parent_get_log(__e); }
    _ => {}
}
    }

    fn _s_Child_event_b(&mut self, __e: &HSMForwardFrameEvent) {
self.log.push("Child:event_b_forward".to_string());
self._state_Parent(__e);
    }

    fn _s_Child_get_log(&mut self, __e: &HSMForwardFrameEvent) {
self.log.clone();
    }

    fn _s_Child_event_a(&mut self, __e: &HSMForwardFrameEvent) {
self.log.push("Child:event_a".to_string());
    }

    fn _s_Parent_event_b(&mut self, __e: &HSMForwardFrameEvent) {
self.log.push("Parent:event_b".to_string());
    }

    fn _s_Parent_get_log(&mut self, __e: &HSMForwardFrameEvent) {
self.log.clone();
    }

    fn _s_Parent_event_a(&mut self, __e: &HSMForwardFrameEvent) {
self.log.push("Parent:event_a".to_string());
    }
}


fn main() {
    println!("=== Test 08: HSM Forward ===");
    let mut s = HSMForward::new();

    // event_a should be handled by Child (no forward)
    s.event_a();
    let log = s.get_log();
    assert!(log.contains(&"Child:event_a".to_string()), "Expected 'Child:event_a' in log, got {:?}", log);
    println!("After event_a: {:?}", log);

    // event_b should forward to Parent
    s.event_b();
    let log = s.get_log();
    assert!(log.contains(&"Child:event_b_forward".to_string()), "Expected 'Child:event_b_forward' in log, got {:?}", log);
    assert!(log.contains(&"Parent:event_b".to_string()), "Expected 'Parent:event_b' in log (forwarded), got {:?}", log);
    println!("After event_b (forwarded): {:?}", log);

    println!("PASS: HSM forward works correctly");
}
